io/src/real_sense_2_grabber.cpp

```diff
@@ -0,0 +1,352 @@
+#include <pcl/common/time.h>
+
+
+namespace pcl
+{
+  using namespace io;
+
+  RealSense2Grabber::RealSense2Grabber ( const std::string& file_name_or_serial_number, const bool repeat_playback )
+    : signal_PointXYZ ( createSignal<signal_librealsense_PointXYZ> () )
+    , signal_PointXYZI ( createSignal<signal_librealsense_PointXYZI> () )
+    , signal_PointXYZRGB ( createSignal<signal_librealsense_PointXYZRGB> () )
+    , signal_PointXYZRGBA ( createSignal<signal_librealsense_PointXYZRGBA> () )
+    , file_name_or_serial_number_ ( file_name_or_serial_number )
+    , repeat_playback_ ( repeat_playback )
+    , quit_ ( false )
+    , running_ ( false )
+    , fps_ ( 0 )
+    , device_width_ ( 424 )
+    , device_height_ ( 240 )
+    , target_fps_ ( 30 )
+  {
+  }
+
+  RealSense2Grabber::~RealSense2Grabber ()
+  {
+    try
+    {
+      stop ( );
+
+      disconnect_all_slots<signal_librealsense_PointXYZ> ( );
+      disconnect_all_slots<signal_librealsense_PointXYZI> ( );
+      disconnect_all_slots<signal_librealsense_PointXYZRGB> ( );
+      disconnect_all_slots<signal_librealsense_PointXYZRGBA> ( );
+    } 
+    catch ( ... ) {}
+  }
+
+  void
+  RealSense2Grabber::start ()
+  {
+    if (isRunning ())
+      return;
+
+    // need at least one signal
+    if (signal_PointXYZ->num_slots () == 0 &&
+      signal_PointXYZI->num_slots () == 0 &&
+      signal_PointXYZRGB->num_slots () == 0 &&
+      signal_PointXYZRGBA->num_slots () == 0)
+      return;
+
+    running_ = true;
+    quit_ = false;
+
+    rs2::config cfg;
+
+    // capture from file
+    if (file_name_or_serial_number_.rfind ( ".bag" ) == file_name_or_serial_number_.length () - 4)
+    {
+      cfg.enable_device_from_file ( file_name_or_serial_number_, repeat_playback_ );
+    }
+    else
+    {
+      if (!file_name_or_serial_number_.empty ())
+        cfg.enable_device ( file_name_or_serial_number_ );
+
+      if (signal_PointXYZRGB->num_slots () > 0 || signal_PointXYZRGBA->num_slots () > 0)
+      {
+        cfg.enable_stream ( RS2_STREAM_COLOR, device_width_, device_height_, RS2_FORMAT_RGB8, target_fps_ );
+      }
+
+      cfg.enable_stream ( RS2_STREAM_DEPTH, device_width_, device_height_, RS2_FORMAT_Z16, target_fps_ );
+
+      if (signal_PointXYZI->num_slots () > 0)
+      {
+        cfg.enable_stream ( RS2_STREAM_INFRARED, device_width_, device_height_, RS2_FORMAT_Y8, target_fps_ );
+      }
+
+    }
+
+    rs2::pipeline_profile prof = pipe_.start ( cfg );
+
+    if ( prof.get_stream ( RS2_STREAM_COLOR ).format ( ) != RS2_FORMAT_RGB8 ||
+      prof.get_stream ( RS2_STREAM_DEPTH ).format ( ) != RS2_FORMAT_Z16 ||
+      prof.get_stream ( RS2_STREAM_INFRARED ).format ( ) != RS2_FORMAT_Y8 )
+      THROW_IO_EXCEPTION ( "This stream type or format not supported." );
+
+    if (signal_PointXYZRGB->num_slots () > 0 || signal_PointXYZRGBA->num_slots () > 0)
+    {
+      assert ( prof.get_stream ( RS2_STREAM_COLOR ).format () == RS2_FORMAT_RGB8 );
```

Comment:
This check is irrelevant because this condition is already tested in line 129. In case it is `false` an exception is thrown.